src/dns_trait.rs

```diff
@@ -0,0 +1,70 @@
+use std::fmt::Display;
+
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+
+use crate::consul;
+
+// convert dnstag type to an enum
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Hash)]
+pub enum DnsType {
+    A,
+    AAAA,
+    CNAME,
+}
+
+// implement FromStr for DnsType
+impl std::str::FromStr for DnsType {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "A" => Ok(DnsType::A),
+            "AAAA" => Ok(DnsType::AAAA),
+            "CNAME" => Ok(DnsType::CNAME),
+            _ => Err(format!("Invalid DNS type: {}", s)),
```

Comment:
It's a pretty minor thing for this specific case, but in general it would be nicer to have a tuple-type struct representing this error case instead.

```rust
struct UnsupportedRecordType(String);
```